### PR TITLE
Update googleJobStore with google client API

### PR DESCRIPTION
**Pull request by jessebrennan**
This replaces the old boto stuff with google's easy to use [client API].(https://googlecloudplatform.github.io/google-cloud-python/latest/storage/client.html)

Server side encryption should be working too!

~It is possible that some tests may fail (they did once while testing) with a 427 too many requests error for creating and destroying buckets too quickly. We could build in a mechanism to  retry if this happens or ask google to increase our limit (see https://cloud.google.com/compute/docs/api-rate-limits)~ Added retry mechanism.

TODO:

- [x]  Update docs for setting up / using google job store
- [x] Set up Jenkins with new credentials

### CODE REVIEW

**Comment by jessebrennan** [comment]
This failed because we need to give Jenkins credentials.

**Comment by jessebrennan** [comment]
Hopefully now any new instance to spin up will have to new key and this will avoid previous authentication issues.

Jenkins test this please.

**Comment by jessebrennan** [comment]
Well it only took forever to test, but I think this is ready for review @ejacox

File: src/toil/jobStores/googleJobStore.py [hunk 1]
```diff
@@ -236,111 +176,104 @@ def writeFile(self, localFilePath, jobStoreID=None):
     @contextmanager
     def writeFileStream(self, jobStoreID=None):
         fileID = self._newID(isFile=True, jobStoreID=jobStoreID)
-        key = self._newKey(fileID)
-        with self._uploadStream(key, update=False) as writable:
-            yield writable, key.name
+        with self._uploadStream(fileID, update=True) as writable:
```
Comment (ejacox) [inline src/toil/jobStores/googleJobStore.py]: Why did update change to True?

File: src/toil/jobStores/googleJobStore.py [hunk 2]
```diff
@@ -411,144 +329,88 @@ def _newID(isFile=False, jobStoreID=None):
...
+        """
+        blob = self.bucket.blob(bytes(fileName), encryption_key=self.sseKey if encrypt else None)
 
         class UploadPipe(WritablePipe):
             def readFrom(self, readable):
-                headers = store.encryptedHeaders if encrypt else store.headerValues
-                if update:
-                    try:
-                        key.set_contents_from_stream(readable, headers=headers)
-                    except GSDataError:
```
Comment (ejacox) [inline src/toil/jobStores/googleJobStore.py]: Why does boto have a different write function for streams?

File: src/toil/common.py [hunk 3]
```diff
@@ -807,6 +807,7 @@ def getJobStore(cls, locator):
             from toil.jobStores.azureJobStore import AzureJobStore
             return AzureJobStore(rest)
         elif name == 'google':
+            #from toil.jobStores.googleJobStore import GoogleJobStore
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/toil/common.py]: Done

File: src/toil/jobStores/abstractJobStore.py [hunk 4]
```diff
@@ -822,18 +828,7 @@ def writeFileStream(self, jobStoreID=None):
...
-        Call to fileExists(getEmptyFileStoreID(jobStoreID)) will return True.
-
-        :param str jobStoreID: the id of a job, or None. If specified, the file will be associated with
-               that job and when jobStore.delete(job) is called a best effort attempt is made to delete
-               all files written with the given job.jobStoreID
-
-        :return: a jobStoreFileID that references the newly created file and can be used to reference the
-                 file in the future.
-        :rtype: str
-        """
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/toil/jobStores/abstractJobStore.py]: I agree. It looks like this happened in e77b2dede610e3600ca6d10a5b3d823010a4f150 actually, but I'm putting it back.

File: src/toil/jobStores/googleJobStore.py [hunk 5]
```diff
@@ -87,78 +64,59 @@ def __init__(self, locator):
...
         self.bucketName = namePrefix+"--toil"
         log.debug("Instantiating google jobStore with name: %s", self.bucketName)
-        self.gsBucketURL = bytes("gs://"+self.bucketName)
-        self._headerValues = {"x-goog-project-id": bytes(projectID)} if projectID else {}
-        self._encryptedHeaders = self.headerValues
 
-        import gcs_oauth2_boto_plugin # needed to import authentication handler
-        self.uri = boto.storage_uri(self.gsBucketURL, GOOGLE_STORAGE)
-        self.files = None
+        import gcs_oauth2_boto_plugin  # needed to import authentication handler
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/toil/jobStores/googleJobStore.py]: Nope nice catch.

File: src/toil/jobStores/googleJobStore.py [hunk 6]
```diff
@@ -40,43 +39,21 @@
...
-from toil.jobStores.aws.utils import SDBHelper
 log = logging.getLogger(__name__)
 
 GOOGLE_STORAGE = 'gs'
 
 
 # TODO
 # - Update gcs_oauth2_boto_plugin.
-# - Check consistency.
 # - update GCE instructions
```
[the remark this answers is not in the document]
Comment (jessebrennan) [reply src/toil/jobStores/googleJobStore.py]: Done.

**Comment by jessebrennan** [comment]
Jenkins test this please

**Comment by ejacox** [comment]
jenkins, test this please

**Comment by jessebrennan** [comment]
CWL conformance test is timing out, but this should be fixed with the latest commit (see #2016).

The other failing test is `toil/src/toil/test/src/fileStoreTest.py::CachingFileStoreTestWithGoogleJobStore::testCacheEvictionFailCase `

Which fails here regularly but also has a flakey history (see #2018).

@ejacox @DailyDreaming This *could* be merged (assuming the cwl test fix works), but then testCacheEvictionFailCase may be failing regularly on master. It may make more sense to wait for me to finish #2018.

**Comment by ejacox** [comment]
Jenkins, test this please

File: src/toil/jobStores/googleJobStore.py (hunk 1 above)
Comment (jessebrennan) [reply src/toil/jobStores/googleJobStore.py]: I can't remember exactly, but I have a feeling tests will fail once I switch it back. Let's see what happens.

**Comment by jessebrennan** [comment]
jenkins test this please

**Comment by jessebrennan** [comment]
jenkins, test this please

**Comment by jessebrennan** [comment]
Looks like google can throw some other kinds of errors that we should retry on:

http://jenkins.cgcloud.info/job/toil-pull-requests/3575/testReport/junit/src.toil.test.sort.sortTest/SortTest/testGoogleSingle/

**Comment by ejacox** [comment]
Jenkins, test this please.

**Comment by jessebrennan** [comment]
@ejacox I think this is ready. I'm pretty sure any left-over bucket issues should be fixed at this point. I am waiting to clean up left-over buckets until no Jenkins builds are running, then I will do more testing to be sure.

**Pull request merged**